src/goto-programs/replace_calls.cpp

```diff
@@ -0,0 +1,158 @@
+#include <util/invariant.h>
+#include <util/irep.h>
+#include <util/string_utils.h>
+#include <util/suffix.h>
+
+/// Replace function calls with calls to other functions
+/// \param goto_model: goto model to modify
+/// \param replacement_list: list of strings, with each string f:g denoting a
+///   mapping between functions names; a mapping f -> g indicates that calls to
+///   f should be replaced by calls to g
+void replace_callst::operator()(
+  goto_modelt &goto_model,
+  const replacement_listt &replacement_list) const
+{
+  replacement_mapt replacement_map = parse_replacement_list(replacement_list);
+  (*this)(goto_model, replacement_map);
+}
+
+/// Replace function calls with calls to other functions
+/// \param goto_model: goto model to modify
+/// \param replacement_map: mapping between function names; a mapping f -> g
+///   indicates that calls to f should be replaced by calls to g
+void replace_callst::operator()(
+  goto_modelt &goto_model,
+  const replacement_mapt &replacement_map) const
+{
+  check_replacement_map(replacement_map);
+
+  const namespacet ns(goto_model.symbol_table);
+  goto_functionst &goto_functions = goto_model.goto_functions;
+
+  for(auto &p : goto_functions.function_map)
+  {
+    goto_functionst::goto_functiont &goto_function = p.second;
+    goto_programt &goto_program = goto_function.body;
+
+    (*this)(goto_program, goto_functions, ns, replacement_map);
+  }
+}
+
+void replace_callst::operator()(
+  goto_programt &goto_program,
+  const goto_functionst &goto_functions,
+  const namespacet &ns,
+  const replacement_mapt &replacement_map) const
+{
+  Forall_goto_program_instructions(it, goto_program)
+  {
+    goto_programt::instructiont &ins = *it;
+
+    if(!ins.is_function_call())
+      continue;
+
+    code_function_callt &cfc = to_code_function_call(ins.code);
+    exprt &function = cfc.function();
+
+    PRECONDITION(function.id() == ID_symbol);
+
+    symbol_exprt &se = to_symbol_expr(function);
+    const irep_idt &id = se.get_identifier();
+
+    auto f_it1 = goto_functions.function_map.find(id);
+
+    INVARIANT(
+      f_it1 != goto_functions.function_map.end(),
+      "Called functions need to be present");
+
+    replacement_mapt::const_iterator r_it = replacement_map.find(id);
+
+    if(r_it == replacement_map.end())
+      continue;
+
+    const irep_idt &new_id = r_it->second;
+
+    auto f_it2 = goto_functions.function_map.find(new_id);
+
+    if(f_it2 == goto_functions.function_map.end())
+      throw "Replacement function " + id2string(new_id) +
+        " needs to be present";
+
+    const goto_functionst::goto_functiont &cur_goto_function = f_it1->second;
+    const goto_functionst::goto_functiont &new_goto_function = f_it2->second;
+
+    if(!base_type_eq(cur_goto_function.type, new_goto_function.type, ns))
+      throw "Functions " + id2string(id) + " and " + id2string(new_id) +
+        " are not type-compatible";
+
+    // check that returns have not been removed
```

Comment:
OK; so function pointer removal but not return removal?  This fits with goto-instrument but not CBMC.  It seems like the kind of thing that should be documented.

Reply:
Yes, if it would be used directly in cbmc it would need to be inserted after the function pointer removal and before the return removal.